src/cascadia/QueryExtension/GithubCopilotLLMProvider.cpp

```diff
@@ -0,0 +1,311 @@
+    }
+
+    void GithubCopilotLLMProvider::SetAuthentication(const Windows::Foundation::Collections::ValueSet& authValues)
+    {
+        _httpClient = winrt::Windows::Web::Http::HttpClient{};
+        _httpClient.DefaultRequestHeaders().Accept().TryParseAdd(L"application/json");
+        _httpClient.DefaultRequestHeaders().Append(L"Copilot-Integration-Id", L"windows-terminal-chat");
+
+        const auto url = unbox_value_or<hstring>(authValues.TryLookup(L"url").try_as<IPropertyValue>(), L"");
+        _authToken = unbox_value_or<hstring>(authValues.TryLookup(L"access_token").try_as<IPropertyValue>(), L"");
+        _refreshToken = unbox_value_or<hstring>(authValues.TryLookup(L"refresh_token").try_as<IPropertyValue>(), L"");
+
+        if (!url.empty())
+        {
+            const Windows::Foundation::Uri parsedUrl{ url };
+            const auto randomStateString = unbox_value_or<hstring>(authValues.TryLookup(L"state").try_as<IPropertyValue>(), L"");
+            // only handle this if the state strings match
+            if (randomStateString == parsedUrl.QueryParsed().GetFirstValueByName(L"state"))
+            {
+                // we got a valid URL, fire off the URL auth flow
+                _completeAuthWithUrl(parsedUrl);
+            }
+        }
+        else if (!_authToken.empty() && !_refreshToken.empty())
+        {
+            // we got tokens, use them
+            _httpClient.DefaultRequestHeaders().Authorization(WWH::Headers::HttpCredentialsHeaderValue{ L"Bearer", _authToken });
+            _obtainUsernameAndRefreshTokensIfNeeded();
+        }
+    }
+
+    winrt::fire_and_forget GithubCopilotLLMProvider::_obtainUsernameAndRefreshTokensIfNeeded()
+    {
+        WWH::HttpRequestMessage request{ WWH::HttpMethod::Get(), Uri{ L"https://api.github.com/user" } };
+        request.Headers().Accept().TryParseAdd(L"application/json");
+        request.Headers().UserAgent().TryParseAdd(L"Windows Terminal");
+
+        co_await winrt::resume_background();
+
+        bool refreshAttempted{ false };
+
+        do
+        {
+            try
+            {
+                const auto response = _httpClient.SendRequestAsync(request).get();
+                const auto string{ response.Content().ReadAsStringAsync().get() };
+                const auto jsonResult{ WDJ::JsonObject::Parse(string) };
+                const auto brandingData{ get_self<GithubCopilotBranding>(_BrandingData) };
+                brandingData->QueryMetaData(jsonResult.GetNamedString(L"login"));
+                break;
+            }
+            catch (...)
+            {
+                // unknown failure, try refreshing the auth token if we haven't already
+                if (!refreshAttempted)
+                {
+                    _refreshAuthTokens();
+                    refreshAttempted = true;
+                }
+                else
+                {
+                    break;
+                }
+            }
+        } while (refreshAttempted);
```

Comment:
This will always we be true.

Reply:
It won't be true if the query succeeds on first pass, meaning we won't try to refresh the tokens (since they worked).

Reply:
There's a break statement in the try clause so if the first pass is successful, it'll exit right after. You could consider changing the code to look like this:
```
for (bool refreshAttempted = false;;) {
    try {
        break; // success -> exit
    } catch(...) {
    }
    
    if (refreshAttempted) {
        break; // already refreshed -> exit
    }
    
    _refreshAuthTokens();
}
```
...or some similar form.